### PR TITLE
doc: Improve documentation of rpcallowip

### LINKED ISSUE #21070: RPC: conf var rpcallowip=::/0 stopped working when upgrading to 0.21
**TL;DR**: When upgrading from `Bitcoin Core 0.20.1` to `Bitcoin Core 0.21.0`, a downstream application using the HTTP JSON-RPC API started getting HTTP 403 "Forbidden".

Turns out it is because I have `rpcallowip=::/0` in `bitcoin.conf` which used to work but does not anymore.
As a workaround, `rpcallowip=0.0.0.0/0` works.

-----

I run bitcoind in docker container1, and some application using the HTTP JSON-RPC API of bitcoind is in docker container2. They are on the same "docker network".

In `bitcoin.conf` I have:
```
rpcauth=<...>
rpcbind=0.0.0.0:8332
rpcallowip=::/0
```
(bitcoind RPC is not actually exposed to the public internet; it is up to the docker layer and above to ensure this.)

The application in container2 is connecting through HTTP accessing an URL such as `http://user:pass@172.19.1.10:8332/`.

This setup used to work when running `Bitcoin Core 0.20.1` in container1.

It no longer works with `Bitcoin Core 0.21.0`.

-----

If I change my `bitcoin.conf` as below, it starts working again using 0.21:
```
rpcauth=<...>
rpcbind=0.0.0.0:8332
rpcallowip=0.0.0.0/0
```

So `rpcallowip=::/0` used to work but does not anymore; `rpcallowip=0.0.0.0/0` works.

-----

Easy way to test:
`curl --data-binary '{"jsonrpc":"1.0","id":"curltext","method":"getblockchaininfo","params":[]}' -H 'content-type:text/plain;' http://user:pass@172.19.1.10:8332/`

### PR DESCRIPTION
**Pull request by willcl-ark**
Closes #21070

v21.0 introduced a behaviour changed noted in #21070 where using a config value `rpcallowip=::0` no longer also permitted ipv4 ip addresses.

The rpc_bind.py functional test covers this new behaviour already by checking that the list of bind addresses exactly matches what is expected so this commit only updates the documentation.

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -596,7 +596,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-blockversion=<n>", "Override block version to test forking scenarios", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::BLOCK_CREATION);
 
     argsman.AddArg("-rest", strprintf("Accept public REST requests (default: %u)", DEFAULT_REST_ENABLE), ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
-    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
+    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24), all ipv4 (0.0.0.0/0) or all ipv6 (::/0). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```
Comment (pinheadmz) [inline src/init.cpp:599]: nit

```suggestion
    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0), a network/CIDR (e.g. 1.2.3.4/24), all ipv4 (0.0.0.0/0), or all ipv6 (::/0). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```

**Review by pinheadmz — APPROVED** [review]
utACK

File: src/init.cpp (hunk 1 above)
Comment (willcl-ark) [reply src/init.cpp:599]: Thanks, I did write it like this intentially as I felt that the final 3 are all variants of network/CIDR, but writing them as a sub-list didn't read well to me...

Thinking more about it I think they should be called wildcard addresses, so perhaps we go with:

```cpp
    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0), a network/CIDR (e.g. 1.2.3.4/24), or wildcard addresses: all ipv4 only (0.0.0.0/0), all ipv6 only (::/0). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```
    
    What do you think?

Comment (pinheadmz) [reply src/init.cpp:599]: Sure ;-)

**Review by brunoerg — COMMENTED** [review]
Concept ACK

File: src/init.cpp (hunk 1 above)
Comment (ekzyis) [reply src/init.cpp:599]: nit:

I would use "or" after "a network/netmask (e.g. 1.2.3.4/255.255.255.0)" now:

```diff
- argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0), a network/CIDR (e.g. 1.2.3.4/24, 0.0.0.0/0 (all ipv4) or ::/0 (all ipv6)). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
+ argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24, 0.0.0.0/0 (all ipv4) or ::/0 (all ipv6)). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);

**Review by ekzyis — APPROVED** [review]

**Comment by willcl-ark** [comment]
> There hasn't been much activity lately. What is the status here?

Happy to close this, if we also close the corresponding issue #21070. But I think it's preferable to keep this open, and merge it to close the issue (and improve the docs).

File: src/init.cpp [hunk 2]
```diff
@@ -610,7 +610,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-blockversion=<n>", "Override block version to test forking scenarios", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::BLOCK_CREATION);
 
     argsman.AddArg("-rest", strprintf("Accept public REST requests (default: %u)", DEFAULT_REST_ENABLE), ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
-    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
+    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0), a network/CIDR (e.g. 1.2.3.4/24), all ipv4 (0.0.0.0/0 ), or all ipv6 (::/0). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```
Comment (pinheadmz) [inline src/init.cpp:613]: nit, extra space ?

```suggestion
    argsman.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0), a network/CIDR (e.g. 1.2.3.4/24), all ipv4 (0.0.0.0/0), or all ipv6 (::/0). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```

Comment (willcl-ark) [reply src/init.cpp:613]: @jonatack took the first. @pinheadmz removed the space. Thanks.

**Review by pinheadmz — APPROVED** [review]
ACK c8e066461b54d745b85411035fcc00a1a4044d76

reviewed code changes, built locally. tested the option with various (in)valid parameters to confirm the docs match the behavior

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK c8e066461b54d745b85411035fcc00a1a4044d76
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmTUwzsACgkQ5+KYS2KJ
yTpDbhAAzRSmfx29YKWZDQ4gY4bCIKF4/PDAwX4DTo3hh7mOpKai7eKx3TPl2RAd
oy42t0dkImgez3gpKKbXzHtapsmJk2GcHCOm0ylC20YrZsX3moaFEErkWT7zBya3
r8+l6p566+g23QzO7UtCAZ+NlNqNpYXz8Si6DkfPhyLtvIgt8LOR7Eu+9YETGyou
lhnocFnWMc5Dg69mFpNVVlHHrTZvnTG3wDXHNRjzz6Elx5gx1/aUAGSJGQXgInSK
F3r8B9mP38jEUnh57c3+acpNZNqSOV+Np32o4jDcwGE0Tpf2pHXsCF5UCzGQEJB/
Bz5qZr6i8Hhr2NLsiDMGQ2UDRGDspKWQumuJWAL5TmD2Wz5IMzaWWL8aL9c4DoSM
nzjXdzF+1gTvXucZHXYm6MFELROODqcNKbnVbd+BW7fC3BZsteqy9+EcobMjzVbt
v3RpGF0msIxc1nb4/kjecrNBLug3vyw9NjE9th+0cRc+sWJByMqgwq3Mo5TMoRvL
mYdsEsDxIIQiRQ3OE73LiQpWZviftQriGDjI9ls9csqBEkYSr572TY+h9H0qFv67
zyKxzresThm1uew1jXDIzIdzA3Rg/RpZllu8GzwhY0Wg4VKgpi7qykkrY8HDywxN
DQ2lleXyWXy2FCsSUzjbbZmUYftbnSRx8e6mTKtt5zrEV7dvmmI=
=d+m3
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Review by jonatack — COMMENTED** [review]
ACK c8e066461b54d745b85411035fcc00a1a4044d76

Non-blockers, consider changing the commit message to something like `doc: update -rpcallowip config option help documentation` (it's not an RPC help), and (i.e., for the future) using `PR 21070` rather than `#21070` so GitHub doesn't add mentions in that issue on every push.

**Pull request merged**